<p align="center">
   <a>
    <img src="https://readme-typing-svg.demolab.com?font=Georgia&size=42&duration=2000&pause=100&multiline=true&width=1000&height=200&lines=Yacine+BERKANI+%7C+Anass+EL+ADI;Master+2+Big+Data+Université+Paris+8+;Projet+IA+Générative" alt="Typing SVG" />
</a>
</p>

# 1. Configuration de l'environnement :

In [1]:
pip install transformers torch streamlit

Note: you may need to restart the kernel to use updated packages.


In [1]:
# Importation des classes nécessaires depuis la bibliothèque `transformers`
from transformers import AutoTokenizer, AutoModelForCausalLM

# Définition de la fonction `telecharger_modeles` qui télécharge les modèles et leurs tokenizers
def telecharger_modeles():
    # Exemple pour GPT-2
    
    # Nom du modèle GPT-2 à télécharger
    model_name_gpt2 = "gpt2"  
    
    # Téléchargement du tokenizer associé au modèle GPT-2
    tokenizer_gpt2 = AutoTokenizer.from_pretrained(model_name_gpt2)
    
    # Téléchargement du modèle GPT-2 lui-même
    model_gpt2 = AutoModelForCausalLM.from_pretrained(model_name_gpt2)
    
    # Affichage d'un message de confirmation pour GPT-2
    print("GPT-2 téléchargé avec succès !")

    # Exemple pour GPT-Neo
    
    # Nom du modèle GPT-Neo à télécharger (ici, la version 125M de paramètres)
    model_name_gptneo = "EleutherAI/gpt-neo-125M"  # vous pouvez choisir une autre taille
    
    # Téléchargement du tokenizer associé au modèle GPT-Neo
    tokenizer_gptneo = AutoTokenizer.from_pretrained(model_name_gptneo)
    
    # Téléchargement du modèle GPT-Neo lui-même
    model_gptneo = AutoModelForCausalLM.from_pretrained(model_name_gptneo)
    
    # Affichage d'un message de confirmation pour GPT-Neo
    print("GPT-Neo téléchargé avec succès !")

# Vérification si le script est exécuté directement (et non importé comme module)
if __name__ == "__main__":
    # Appel de la fonction `telecharger_modeles` pour lancer le téléchargement
    telecharger_modeles()

GPT-2 téléchargé avec succès !
GPT-Neo téléchargé avec succès !


In [3]:
# Importation de la bibliothèque PyTorch, utilisée pour le calcul tensoriel et les réseaux de neurones
import torch

# Importation de la bibliothèque Streamlit, utilisée pour créer des applications web interactives
import streamlit

# Importation des classes nécessaires depuis la bibliothèque `transformers` pour charger les modèles de langage et leurs tokenizers
from transformers import AutoTokenizer, AutoModelForCausalLM

# Affichage d'un message confirmant que les bibliothèques ont été importées avec succès
print("Bibliothèques importées avec succès !")

# (Optionnel) Test sur GPT-2

# Chargement du tokenizer associé au modèle GPT-2
tokenizer = AutoTokenizer.from_pretrained("gpt2")

# Chargement du modèle GPT-2 lui-même
model = AutoModelForCausalLM.from_pretrained("gpt2")

# Affichage d'un message confirmant que GPT-2 a été chargé sans problème
print("GPT-2 chargé sans problème !")

Bibliothèques importées avec succès !
GPT-2 chargé sans problème !


# 2. Génération de texte :

In [13]:
pip install evaluate

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 84.0/84.0 kB 415.1 kB/s eta 0:00:00a 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [60]:
pip install sacrebleu

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.8/51.8 kB 527.7 kB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 104.1/104.1 kB 466.3 kB/s eta 0:00:00a 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [112]:
# Importe la bibliothèque PyTorch, utilisée pour le calcul tensoriel et les réseaux de neurones
import torch

# Importe la classe AutoTokenizer pour charger des tokenizers pré-entraînés
from transformers import AutoTokenizer

# Importe la classe AutoModelForCausalLM pour charger des modèles de langage causal pré-entraînés
from transformers import AutoModelForCausalLM

# Importe la fonction corpus_bleu de la bibliothèque sacrebleu pour calculer le score BLEU
from sacrebleu import corpus_bleu

# Importe la bibliothèque evaluate pour évaluer les modèles de traitement du langage naturel
import evaluate

In [118]:
# Définition de la fonction `load_model` qui charge un tokenizer et un modèle de langage
def load_model(model_name):
    """
    Charge le tokenizer et le modèle correspondants.
    """
    # Charge le tokenizer associé au modèle spécifié par `model_name`
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    
    # Charge le modèle de langage causal (comme GPT-2 ou GPT-Neo) spécifié par `model_name`
    model = AutoModelForCausalLM.from_pretrained(model_name)
    
    # Retourne le tokenizer et le modèle chargés
    return tokenizer, model

# Exemple d'utilisation de la fonction `load_model` pour charger GPT-2 et GPT-Neo

# Charge le tokenizer et le modèle pour GPT-2
tokenizer_gpt2, model_gpt2 = load_model("gpt2")

# Charge le tokenizer et le modèle pour GPT-Neo (version 125M de paramètres)
tokenizer_gptneo, model_gptneo = load_model("EleutherAI/gpt-neo-125M")

In [120]:
# Importation des classes nécessaires depuis la bibliothèque `transformers`
from transformers import AutoTokenizer, AutoModelForCausalLM
# Importation de PyTorch pour gérer les tenseurs et les calculs
import torch

# Définition de la fonction `generate_text` pour générer du texte à partir d'un prompt
def generate_text(tokenizer, model, prompt, max_length=100, temperature=1.0, top_p=0.9):
    """
    Génère du texte à partir d'un prompt en utilisant un modèle de langage.
    """
    # Tokenisation du prompt : conversion du texte en entrée en un format tensoriel (compréhensible par le modèle)
    inputs = tokenizer(prompt, return_tensors="pt")
    
    # Génération du texte
    with torch.no_grad():  # Désactive le calcul des gradients pour économiser de la mémoire
        output_ids = model.generate(
            **inputs,  # Passe les entrées tokenisées au modèle
            max_length=max_length,  # Longueur maximale du texte généré
            do_sample=True,         # Active l'échantillonnage pour plus de créativité
            top_p=top_p,            # Contrôle la diversité des mots générés (échantillonnage par noyau)
            temperature=temperature,  # Contrôle la créativité (plus élevé = plus aléatoire)
            num_return_sequences=1   # Nombre de séquences à générer
        )
    
    # Décodage des IDs générés en texte lisible
    generated_text = tokenizer.decode(output_ids[0], skip_special_tokens=True)
    return generated_text  # Retourne le texte généré

# Définition de la fonction `generer_poemes` pour générer des poèmes avec GPT-2 et GPT-Neo
def generer_poemes(prompt, max_length=100, temperature=1.0, top_p=0.9):
    """
    Génère un poème avec GPT-2 et GPT-Neo en utilisant un prompt optimisé.
    """
    # Chargement des modèles et tokenizers pour GPT-2 et GPT-Neo
    tokenizer_gpt2, model_gpt2 = load_model("gpt2")
    tokenizer_gptneo, model_gptneo = load_model("EleutherAI/gpt-neo-125M")
    
    # Génération des poèmes avec GPT-2 et GPT-Neo
    poeme_gpt2 = generate_text(tokenizer_gpt2, model_gpt2, prompt, max_length, temperature, top_p)
    poeme_gptneo = generate_text(tokenizer_gptneo, model_gptneo, prompt, max_length, temperature, top_p)
    
    # Retourne les deux poèmes générés
    return poeme_gpt2, poeme_gptneo


In [128]:
# Exemple d'utilisation
if __name__ == "__main__":
    # Prompt optimisé en anglais pour la génération de poèmes
    prompt_user = (
        "Write a romantic free verse poem on the theme of love and freedom. "
        "The poem should evoke deep emotions and use poetic language."
    )
    
    # Paramètres de génération
    max_length = 120  # Longueur maximale du poème
    temperature = 1.1  # Créativité accrue (plus élevé = plus aléatoire)
    top_p = 0.9       # Diversité des mots (échantillonnage par noyau)
    
    # Génération des poèmes avec GPT-2 et GPT-Neo
    poeme_gpt2, poeme_gptneo = generer_poemes(prompt_user, max_length, temperature, top_p)
    
    # Affichage des résultats
    print("=== Poème généré par GPT-2 ===")
    print(poeme_gpt2)
    print("\n=== Poème généré par GPT-Neo ===")
    print(poeme_gptneo)

Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


=== Poème généré par GPT-2 ===
Write a romantic free verse poem on the theme of love and freedom. The poem should evoke deep emotions and use poetic language. The poem should be simple and straightforward and say good and bad to your partner, and your partner can answer any questions. If you have never heard of any poem that calls for freedom then this will be a perfect poem.

Read a couple-written romance poem to be heard. I am a married woman living in the Midwest of USA.

Love in action! If you've never heard of the love poetry for the two songs on the list, you could try

=== Poème généré par GPT-Neo ===
Write a romantic free verse poem on the theme of love and freedom. The poem should evoke deep emotions and use poetic language. The poem should contain elements of beauty and romance that evoke feelings of inner peace and connection. This poem should be a response to love in the heart of one’s partner, family, or other community.

What will your Valentine look like and why is the r

In [130]:
def compute_bleu_score(text_ref, text_hyp):
    """
    Calcule un score BLEU (simple) entre un texte de référence et un texte hypothèse.
    Le score BLEU est une métrique utilisée pour évaluer la similarité entre deux textes,
    souvent utilisée pour évaluer la qualité des traductions ou des générations de texte.
    """
    # Assurez-vous que les entrées sont des chaînes de caractères
    reference = text_ref.strip()
    hypothesis = text_hyp.strip()

    # sacrebleu attend une liste de références
    # et l'hypothèse comme une seule chaîne de caractères
    bleu = sentence_bleu(hypothesis, [reference])

    # Retourne le score BLEU sous forme de float
    return bleu.score / 100.0  # sacrebleu retourne un score en pourcentage

In [132]:
# Chargement de la métrique ROUGE à l'aide de la bibliothèque `evaluate`
rouge_metric = evaluate.load("rouge")

def compute_rouge_score(text_ref, text_hyp):
    """
    Calcule les scores ROUGE (rouge1, rouge2, rougel, etc.) 
    entre un texte de référence et un texte hypothèse.
    ROUGE est une métrique couramment utilisée pour évaluer la qualité des résumés automatiques
    ou des générations de texte en comparant la similarité avec un texte de référence.
    """
    # Calcul des scores ROUGE en utilisant la métrique chargée
    # `predictions` : liste contenant le texte hypothèse (généré par le modèle)
    # `references` : liste contenant le texte de référence (texte de comparaison)
    results = rouge_metric.compute(
        predictions=[text_hyp],  # Texte hypothèse (généré)
        references=[text_ref]    # Texte de référence
    )
    
    # Retourne un dictionnaire contenant les scores ROUGE
    # Les clés du dictionnaire sont : 'rouge1', 'rouge2', 'rougeL', etc.
    return results

In [153]:
def comparer_textes(text_gpt2, text_gptneo):
    """
    Compare text_gpt2 et text_gptneo en calculant les scores BLEU et ROUGE.
    On choisit, par exemple, text_gpt2 comme 'référence' et text_gptneo comme 'hypothèse'.
    """
    # Appelle la fonction pour calculer le score BLEU entre text_gpt2 (référence) et text_gptneo (hypothèse)
    bleu = compute_bleu_score(text_gpt2, text_gptneo)
    
    # Appelle la fonction pour calculer le score ROUGE entre text_gpt2 (référence) et text_gptneo (hypothèse)
    rouge = compute_rouge_score(text_gpt2, text_gptneo)
    
    # Retourne un dictionnaire contenant les scores BLEU et ROUGE
    return {
        "BLEU": bleu,
        "ROUGE": rouge
    }

In [149]:
# Vérifie si le script est exécuté directement (et non importé comme module)
if __name__ == "__main__":
    
    # Comparaison des deux poèmes en calculant les scores BLEU et ROUGE
    scores = comparer_textes(poeme_gpt2, poeme_gptneo)
    
    # Affichage des résultats de la comparaison
    print("Résultats de la comparaison :")
    for k, v in scores.items():
        print(f"{k} = {v}")

Résultats de la comparaison :
BLEU = 0.30851415778293173
ROUGE = {'rouge1': 0.5333333333333333, 'rouge2': 0.2980769230769231, 'rougeL': 0.38095238095238093, 'rougeLsum': 0.44761904761904764}


## Interpréter les scores obtenus

## Score BLEU : 0.3085

- **Le score BLEU de 0.3085 indique une similarité modérée entre les textes générés par GPT-2 et GPT-Neo en termes de précision des n-grammes. Ce score suggère que les textes ne sont pas très similaires en termes de structure et de cohérence des phrases.**

## Score ROUGE :

- **Les scores ROUGE (ROUGE-1 : 0.5333, ROUGE-2 : 0.2981, ROUGE-L : 0.3810, ROUGE-Lsum : 0.4476) montrent une similarité modérée à élevée en termes de mots individuels et de séquences de mots entre les textes générés par les deux modèles. Cela indique que plus de la moitié des mots individuels sont communs, mais la similarité diminue pour les paires de mots et les séquences plus longues.**






In [ ]:
!streamlit run app.py


  You can now view your Streamlit app in your browser.

  Local URL: http://localhost:8501
  Network URL: http://192.168.1.194:8501

/opt/anaconda3/lib/python3.12/site-packages/streamlit/util.py:227: RuntimeWarning: coroutine 'expire_cache' was never awaited
  pass
2025-01-16 23:19:53.974 Examining the path of torch.classes raised: Tried to instantiate class '__path__._path', but it does not exist! Ensure that it is registered via torch::class_
2025-01-16 23:19:54.102 Examining the path of torch.classes raised: Tried to instantiate class '__path__._path', but it does not exist! Ensure that it is registered via torch::class_
2025-01-16 23:19:54.229 Examining the path of torch.classes raised: Tried to instantiate class '__path__._path', but it does not exist! Ensure that it is registered via torch::class_
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
huggingface/tokenizers: The current proc